awx/main/managers.py

```diff
@@ -271,3 +273,19 @@ def capacity_values(self, qs=None, tasks=None, breakdown=False, graph=None):
             else:
                 logger.error('Programming error, %s not in ["running", "waiting"]', t.log_format)
         return graph
+
+
+class ExecutionEnvironmentManager(models.Manager):
+    '''Allows caching for some global functions'''
+
+    def get_default_execution_environment(self):
+        if settings.DEFAULT_EXECUTION_ENVIRONMENT is not None:
+            return settings.DEFAULT_EXECUTION_ENVIRONMENT
+        installed_default = self.model.objects.filter(
+            image__in=[ee['image'] for ee in settings.GLOBAL_JOB_EXECUTION_ENVIRONMENTS], organization=None, managed=False
+        ).first()
+        if installed_default:
+            return installed_default
+        return self.model.objects.filter(organization=None, managed=False).first()
+
+    default_execution_environment = cached_property(get_default_execution_environment, name='default_execution_environment')
```

Comment:
@jbradberry Picking up the conversation from:

https://github.com/ansible/awx/pull/11034#issuecomment-914723421

This is the pattern I would prefer to use for these static, global, objects. We can use the cached property with abandon, because it will never accumulate queries. This is good if the method we're writing is going to be used incidentally deep inside the task manager.

Reply:
I think that using `cached_property` is going to be too "sticky".  What happens when the sysadmin changes `settings.DEFAULT_EXECUTION_ENVIRONMENT` or makes other changes that would affect which EE is considered the global default?

Also, this cache is just an attribute stashed on an instantiation of the custom manager.  This value isn't going to be shared amongst different server processes, and even within a single process may go out of scope and need to be re-queried unexpectedly.

I'm not sure which of these potential problems would be the bigger annoyance, or whether they would wind up mitigating each other in practice.

Reply:
So far, the only other caching option I've thought about is stashing it in the view context (so list views won't re-query same object 25x). That's going to get messy across multiple view, and it's not quite as aggressive as I'd like. If the task manager grabbed this once for all pending jobs, I'd like all but 1 to use a cache.

It might be best to allow it fall back to the actual cache via the settings. But I don't really know what the ForeignKey mechanics are for that, and the `settings.DEFAULT_EXECUTION_ENVIRONMENT` is only one of the code paths that it might ultimately end up on.

Reply:
We could use Django's caching system directly, and stash the pk of the resolved global EE in whatever our backing store is.  A signal handler on EEs and on the `settings.DEFAULT_EXECUTION_ENVIRONMENT` value could then invalidate the cache when something changes.